### ISSUE TITLE
Getting "ScErrIteratorClosed"?

### ISSUE THREAD

**Issue by malx122** [Starcounter/Home#98]
This now happens all the time on our intranet! 

```
-> Starcounter.DbException: ScErrIteratorClosed (SCERR4139): A interator operation failed because the iterator was closed.
Version: 2.3.0.6147.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR4139.
   at Starcounter.FilterEnumerator.MoveNext() in C:\TeamCity\BuildAgent\work\sc-4222\Level1\src\Starcounter\Mock\Enumerator.cs:line 482
   at Starcounter.Query.Execution.FullTableScan.MoveNext() in C:\TeamCity\BuildAgent\work\sc-4222\Level1\src\Starcounter\Query\Execution\Enumerators\FullTableScan.cs:line 479
   at Starcounter.Query.Execution.Join.NextRightAndLeftEnumerators() in C:\TeamCity\BuildAgent\work\sc-4222\Level1\src\Starcounter\Query\Execution\Enumerators\Join.cs:line 281
   at Starcounter.Query.Execution.Join.MoveNext() in C:\TeamCity\BuildAgent\work\sc-4222\Level1\src\Starcounter\Query\Execution\Enumerators\Join.cs:line 253
   at Concepts.Ring8.Tunity.Todo.<get_SubscribedUsers>d__69.MoveNext()
   at LogHandlers.<>c.<Register>b__0_22(LogEntry le)
   at Colab.Common.LogManager.CallCallback(LogEntry newEntry)
   at Colab.Common.LogManager.AddEntry(TunityUser modifier, String name, String propertyName, Something target, Something data, DateTime time, Boolean simulated)
   at Colab.Common.LogManager.AddEntry(String name, Something target, Something data)
   at Colab.Common.ActionBoardHelper.ApplyColumnData(Todo todo, BoardColumn column)
   at Colab.Common.ActionBoardHelper.AddToColumn(Todo todo, BoardColumn column, Int32 index)
   at Colab.Actions.ActionDetails.set_ColumnCB(String value)
   at lambda_method(Closure , Json , String )
ErrorCode=4139
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR4139
```
The code for the function:
```cs
foreach(TunityUser user in Db.SQL<TunityUser>("SELECT p FROM Concepts.Ring8.Tunity.TunityUser p JOIN Concepts.Ring8.Tunity.TunitySubscriber r WHERE r.WhatIs=p AND r.ToWhat=?", this))
{
    if (user.Active)
        yield return user;
}
```

**Comment by miyconst** [comment]
@malx122 did you do anything, like updating the Starcounter version, or changing database classes? Is it 100% reproducible or crashes sometimes and sometimes not?

**Comment by miyconst** [comment]
Aside from the issue. What is the reason of using the `JOIN` and `yield` operator if the same thing can be done simpler?

First add `User` property to the `TunitySubscriber` class:

```cs
public TunityUser User
{
	get
	{
		return this.WhatIs as TunityUser;
	}
	set
	{
		this.WhatIs = value;
	}
}
```

And then use a single query:

```cs
var activeUsers = Db.SQL<TunityUser>("SELECT s.User FROM Concepts.Ring8.Tunity.TunitySubscriber s WHERE s.ToWhat = ? AND s.User.Active = ?", this, true);
```

Am I missing something?

**Comment by malx122** [comment]
It seems to be 100% reproducible, but I don't really know what causing it. SC version has not been changed, the code itself has not been changed. However the code calling it have probably been changed.

So my question is really, what can trigger this error? What can cause an iterator to close?

**Comment by miyconst** [comment]
If a transaction was closed or aborted by someone, the iterator will be closed as well.

I am not sure if that is the case, but it also might get closed, if you start another iterator. @k-rus please comment on this.

**Comment by malx122** [comment]
Reason for the query is probably that "someone" has made a quick fix (copy/paste), instead of thinking it through.
btw, what about `[SynonymousTo("WhatIs")]` in your example?

**Comment by miyconst** [comment]
The `SynonymousTo` is just a short cut which produces exactly the same code as the custom property I provided above.

We found that explaining `SynonymousTo` is very difficult, while clean C# is obvious for everyone who knows the language. Also `SynonymousTo` throws exception if the actual `WhatIs` value happens to be of other type rather than it's synonym.

That is why `SynonymousTo` will be deprecated soon and is no longer recommended to use.

**Comment by malx122** [comment]
Hm, can a Self.GET('...') trigger the iterator to close?

**Comment by miyconst** [comment]
It probably depends on what is happening in the requested handler.

**Comment by malx122** [comment]
It can be handled in different applications by mapping so I can never control that.

**Comment by malx122** [comment]
Nope, it was probably not that one.

**Comment by malx122** [comment]
Ok, it is database commits that's not allowed.
I did this (but more complicated inside several functions so hard to see):
```
//Someone did something that should be notified to several users
foreach (TunityUser user in action.SubscribedUsers)
{
   var not = new NotificationOfRightType();
   Transaction.Commit();
    ColabX.PublishMessage(ColabAPI.Messages.NotificationAdded, not.DbIDString);   
}
```
PublishMessage is our own pub/sub protocol to send messages between apps with Self.Get and since, in this case, the Notification app needs to be able to get the notification and publish it/send mail e.t.c. it needs to be commited to the db.
Until now we have used the starcounter database hooks instead but they have been really unreliable (under a couple of circumstances they don't get triggered) and thats why I changed it to this.
